oppia/oppia-android
# Fix #1382, #1340, #1341: Show a dialog when preloading assets for offline support

Fix #1382
Fix #1340
Fix #1341

Show a dialog when preloading assets for offline support.

### Test instructions

1. Check out this branch locally
1. Enable local file caching by setting this boolean to true: https://github.com/oppia/oppia-android/blob/develop/utility/src/main/java/org/oppia/util/caching/CachingModule.kt#L11.
1. If the app is already installed on your device/emulator, go into app settings and clear storage (not just cache)
1. Build & install the app
1. Open the app--the dialog should appear

### Expanded description

This does a bunch of refactoring & hacky workarounds to keep asset priming support fully isolated (just a few files now need to be deleted to clean things up). This also:

1. Fixes actual asset downloading (the GCS asset path templates have changed, so priming didn't actually work anymore).
2. Disables audio file caching since we can't play audio when offline, anyway (there's a dialog that prevents this in-player).
3. Fixes #1340 and #1341 by accounting for error cases when trying to play audio. It turns out that playing audio crashes if you didn't have internet access when going into an exploration (the no connectivity dialog only appears if you lose connectivity within a lesson). There's also some issues with existing view model code that wrongly assumed audio couldn't be in a failure state at that point. This has been fixed.
4. Moves the list of topics to cache to be in the same position as the flag enabling/disabling this functionality.

The download experience isn't perfect, but it's meant to be a helper for user study coordinators so that they know when lessons can be brought offline.

Note that the UI aspects of this change are really hacky. This is by design--I didn't want to overcomplicate the solution, and I wanted to keep the priming changes fully isolated to make future cleanup easier.

Finally, no new tests were added. I clarified in StateFragmentTest that the audio edge cases fixed in this PR need to have corresponding tests, but I'm actually not sure offhand how to test that audio's playing. I think this will require additional work. I prefer to push this off to #388, but I will follow up with tests if anyone wants to push back on this. I'm also not inclined to test the priming behavior since that's meant to be temporary, and is turned off except for alpha (at which point I will retest the behavior before uploading it to the play store).

UI screen recording (happy path):
![priming_support](https://user-images.githubusercontent.com/12983742/90085047-dba95180-dccb-11ea-9b3f-920c2f22c614.gif)

UI screen recording (failure):
![priming_failed](https://user-images.githubusercontent.com/12983742/90085103-fe3b6a80-dccb-11ea-8c70-2bf504f91e89.gif)
